### PR TITLE
core(tracehouse): add CPU trace profiler model

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Phase 1 of #8526. Adds support in tracehouse for parsing CPU profiler events and converting them into a format that our pre-existing main-thread-tasks task tree model can understand. This has no impact on Lighthouse or tracehouse *yet* as its not in the default flow.

Followups will...

- Enable main-thread-tasks to use this model and create `attributeableURLs` based on the URLs in profiler data.
- Turn on the profiler trace category by default.

**Related Issues/PRs**
#8526

### CODE REVIEW

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    const map = new Map();
+    for (const node of this._profile.nodes) {
+      map.set(node.id, node);
+    }
+
+    return map;
+  }
+
+  /**
+   * Initialization function to enable O(1) access to nodes by node ID.
```
Comment (patrickhulce) [inline lighthouse-core/lib/tracehouse/cpu-profile-model.js:66]:
```suggestion
   * Initialization function to enable O(1) access to the set of active nodes in the stack by node ID.
```

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+   * Generates the necessary B/E-style trace events for a single transition from stack A to stack B
+   * at the given timestamp.
+   *
+   * Example:
+   *
+   *    timestamp 1234
+   *    previousNodeIds 1,2,3
+   *    currentNodeIds 1,2,4
+   *
+   *    yields [end 3, begin 4]
```
Comment (patrickhulce) [inline lighthouse-core/lib/tracehouse/cpu-profile-model.js:117]:
```suggestion
   *    yields [end 3 at ts 1234, begin 4 at ts 1234]
```

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+ * category into B/E-style trace events that main-thread-tasks.js already knows how to parse into a task tree.
+ *
+ * The CPU profiler measures where time is being spent by sampling the stack (See https://www.jetbrains.com/help/profiler/Profiling_Guidelines__Choosing_the_Right_Profiling_Mode.html
+ * for a generic description of the differences between tracing and sampling).
+ *
+ * A `Profile` event is a record of the stack that was being executed at different sample points in time.
+ * It has a structure like this:
+ *
+ *    nodes: [function A, function B, function C]
+ *    samples: [node at index 2, node at index 1, ...]
```
Comment (patrickhulce) [inline lighthouse-core/lib/tracehouse/cpu-profile-model.js:21]:
```suggestion
 *    samples: [node with id 2, node with id 1, ...]
```

**Comment by patrickhulce** [comment]
most of these lines are tests and trace data I promise @brendankenny 😄

**Comment by patrickhulce** [comment]
friendly ping since there are many more steps to go that are blocked on this :) 

if too much on your plate right now (github loves to assign my PRs to you I guess @brendankenny 😆 ) perhaps someone else might be willing to volunteer?

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+      .filter(/** @return {node is CpuProfile['nodes'][0]} */ node => !!node);
+
+    /** @param {CpuProfile['nodes'][0]} node @return {LH.TraceEvent} */
+    const createEvent = node => ({
+      ts: timestamp,
+      pid: this._profile.pid,
+      tid: this._profile.tid,
+      dur: 0,
+      ph: 'I',
+      name: 'FunctionCall-ProfilerModel',
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:141]: Yes this is a Lighthouse-only trace event name. You would not find it in a raw DevTools trace. 

I modeled it's name on the *real* `FunctionCall` event we already process, but something more obvious that it's fake could be used. 
 

Do any of...

- `FunctionCall-SynthesizedByProfilerModel`
- `FunctionCall-FakedByProfilerModel`
- `ProfilerModelSyntheticFunctionCall`

sound better to you?

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ *
+ * The CPU profiler measures where time is being spent by sampling the stack (See https://www.jetbrains.com/help/profiler/Profiling_Guidelines__Choosing_the_Right_Profiling_Mode.html
+ * for a generic description of the differences between tracing and sampling).
+ *
+ * A `Profile` event is a record of the stack that was being executed at different sample points in time.
+ * It has a structure like this:
+ *
+ *    nodes: [function A, function B, function C]
+ *    samples: [node with id 2, node with id 1, ...]
+ *    timeDeltas: [4125μs since last sample, 121μs since last sample, ...]
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:22]: Yes, if they aren't then it's invalid data. This is checked in the model by https://github.com/GoogleChrome/lighthouse/pull/11072/files#diff-079972702d9ee8e3325ad9e6617563ffR161

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  static collectProfileEvents(traceEvents) {
+    /** @type {Map<string, CpuProfile>} */
+    const profiles = new Map();
+    for (const event of traceEvents) {
+      if (event.name !== 'Profile' && event.name !== 'ProfileChunk') continue;
+      if (typeof event.id !== 'string') continue;
+
+      const cpuProfileArg = (event.args.data && event.args.data.cpuProfile) || {};
+      const timeDeltas =
+        (event.args.data && event.args.data.timeDeltas) || cpuProfileArg.timeDeltas;
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:217]: Kind of. At this point we don't know if it's a `Profile` or `ProfileChunk` and from reading the other trace parsers it appears that it's possible for either to not contain any sample/timeDelta information (and only be about nodes for example). I wasn't able to observe this behavior in modern Chrome, but we kinda need to do this to make typescript happy anyway.

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  }
+
+  /**
+   * Merges the data of all the `ProfileChunk` trace events into a single CpuProfile object for consumption
+   * by `createStartEndEvents()`.
+   *
+   * @param {Array<LH.TraceEvent>} traceEvents
+   * @return {Array<CpuProfile>}
+   */
+  static collectProfileEvents(traceEvents) {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:208]: Sounds good I'll codify up the edge cases mentioned in https://github.com/GoogleChrome/lighthouse/pull/11072#discussion_r468824754 👍

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:141]: Sounds good.  Maybe a comment explaining as well.

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:217]: Added a comment and tests for this case 👍

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:141]: `FunctionCall-SynthesizedByProfilerModel` looks best IMO. I prefer keeping the `FunctionCall` in the name.

File: types/externs.d.ts [hunk 8]
```diff
@@ -233,6 +233,12 @@ declare global {
       [futureProps: string]: any;
     }
 
+    export interface TraceCpuProfile {
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply types/externs.d.ts:236]: I think these are different though?  In actual traces across Chrome versions I saw...

- definition split across `Profile` and `ProfileChunk` that does not guarantee existence of any particular property and fallbacks necessary for profile boundaries
- use of a `parent` numeric ID and not `children` as an array like in the protocol version

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+ *    timeDeltas: [4125μs since last sample, 121μs since last sample, ...]
+ *
+ * Helpful prior art:
+ * @see https://cs.chromium.org/chromium/src/third_party/devtools-frontend/src/front_end/sdk/CPUProfileDataModel.js?sq=package:chromium&g=0&l=42
+ * @see https://github.com/v8/v8/blob/99ca333b0efba3236954b823101315aefeac51ab/tools/profile.js
+ * @see https://github.com/jlfwong/speedscope/blob/9ed1eb192cb7e9dac43a5f25bd101af169dc654a/src/import/chrome.ts#L200
+ */
+
+/**
+ * @typedef CpuProfile
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:31]: I don't think it does, see below.

File: lighthouse-core/lib/tracehouse/cpu-profile-model.js [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+ * @fileoverview
+ *
+ * This model converts the `Profile` and `ProfileChunk` mega trace events from the `disabled-by-default-v8.cpu_profiler`
+ * category into B/E-style trace events that main-thread-tasks.js already knows how to parse into a task tree.
+ *
+ * The CPU profiler measures where time is being spent by sampling the stack (See https://www.jetbrains.com/help/profiler/Profiling_Guidelines__Choosing_the_Right_Profiling_Mode.html
+ * for a generic description of the differences between tracing and sampling).
+ *
+ * A `Profile` event is a record of the stack that was being executed at different sample points in time.
+ * It has a structure like this:
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/lib/tracehouse/cpu-profile-model.js:18]: I can reference this as prior art but I believe it's subtly different and don't want misleading types, see below

File: types/externs.d.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (patrickhulce) [reply types/externs.d.ts:236]: comments to this effect and the fact that it's *not* Crdp.Profiler.Profile sgtm though :)

**Comment by patrickhulce** [comment]
friendly ping @paulirish  :)

File: types/externs.d.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply types/externs.d.ts:236]: yeah sounds good.

wfm

**Pull request merged**